<a href="https://colab.research.google.com/github/adi45-del/Classification-with-Logistic-Regression/blob/main/AI_ML(Task_4).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import  confusion_matrix, precision_score, recall_score, roc_auc_score

df=pd.read_csv("data.csv")
print(df.head())
print(df.info)

print(df.isnull().sum())
df=df.drop(["id","Unnamed: 32"],axis=1)

df["diagnosis"]=df["diagnosis"].map({"B":0,"M":1})
X=df.drop("diagnosis",axis=1)
y=df["diagnosis"]
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)

model=LogisticRegression(max_iter=10000)
model.fit(X_train,y_train)

y_pred = model.predict(X_test)
print(y_pred[:5])

cm=confusion_matrix(y_test,y_pred)
print(cm)

precision=precision_score(y_test,y_pred)
print("precision:",precision)

recall=recall_score(y_test,y_pred)
print("Recall:",recall)

y_prob=model.predict_proba(X_test)[:,1]
roc=roc_auc_score(y_test,y_prob)
print("ROC-AUC:",roc)



         id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness

In [ ]:
from google.colab import files
uploaded = files.upload()

Sigmoid function
Logistic regression uses sigmoid function:

σ(x)=
1+e
−x
1
	​


It converts any number into 0 to 1 probability.

Threshold Tuning

Probability ≥ 0.5 → malignant (1)
Probability < 0.5 → benign (0)

Example:

Probability	Prediction
0.81	1
0.20	0

Changing threshold affects:

Precision
Recall
